### DISCUSSION TITLE
Type-dependent header-format / Query multiple libraries

### DISCUSSION THREAD

**Discussion by choreutes** [papis/papis#717] [Q&A] [answered]
Hi,

is it possible to display different types of documents in different ways in the tui? I might for example want to display the journaltitle for articles but not for books because they don't have one. (Yes, the journaltitle is just empty in that case but I can't get rid of surrounding whitespace if there is some.)

I could, of course, define several libraries (one for articles, one for books, ...) and configure a fitting header-format for each of them. But then I lose the capability to query for all types of documents in one go.

Is there any way to get the best of both worlds? Or do I need to decide for one of the two? Any help is greatly appreciated!

**Comment by alexfikl** [comment, marked as answer]
This isn't possible out of the box at the moment, i.e. all documents get the same `header-format`. However, you can hack around it, e.g. from the top of my mind:

1. **Use the `jinja2` formatter**. One choice that works now is to switch to the fancier `jinja2` formatter, which can use some basic control flow (e.g `{% if journal not in doc %} {{ doc.booktitle }} {% else %} {{ doc.journal }}`). This is a bit cumbersome at the moment because you'd have to turn all the format options to `jinja2`, but #711 should help with that hopefully in time for papis 0.14.

2. **Implement your own formatter plugin**. Besides that, you could also create a custom formatter based on the `PythonFormatter`
https://github.com/papis/papis/blob/c6f54802a62d27f3ea2e04f12ab0fc2f3642c169/papis/format.py#L60
In `Formatter.format` you could catch `header-format` and transform it to `header-format-article` or `header-format-book` and format that instead. For an example of this, you can check out
https://gitlab.com/alexfikl/papis-extras/-/blob/9d84ec70d60d9ac526a5431859bf987ee03f3d48/papis_extras/format.py

**Reply by choreutes** [reply]
Thanks a lot for your very elaborate answer! Those were some really helpful hints. I'll have a look into formatter plugins as you suggested but what I'll probably end up doing is stick to the generic formatter for now until 0.14 hits the floor. Looking forward to that :)
